### PR TITLE
Add `jest-axe`

### PR DESCRIPTION
**Pull request by lluia**
## Summary 💭

Following [this conversation](https://github.com/reakit/reakit/issues/399#issuecomment-630174229), we thought to verify **a11y** violations in tests is a good idea.

This PR adds [`jest-axe`](https://www.npmjs.com/package/jest-axe) and it demonstrates how it could be used within the project.

It extends Jest with a custom matcher:
```js
expect(results).toHaveNoViolations()
```
And suggests to add new util to `reakit-test-utils` → `verify()` given we need to anyway [configure **axe**](https://github.com/nickcolley/jest-axe/issues/92#issuecomment-629809769) when testing components in isolation.

If we're happy with the approach, then I can:
1. Open a new PR bumping `reakit-test-utils` with `verify()`
2. Update on this PR all the existing test suites with `verify()` checks

### CODE REVIEW

File: packages/reakit-test-utils/src/verify.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { configureAxe } from "jest-axe";
+
+export const verify = configureAxe({
```
Comment (diegohaz) [inline packages/reakit-test-utils/src/verify.ts]: Naming it `axe` seems to be the convention. I'd suggest to rename it so (and also rename the file). But let me know if there's any other reason to name it differently.

File: packages/reakit/src/Box/__tests__/Box-test.tsx [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import * as React from "react";
 import { render } from "reakit-test-utils";
+import { verify } from "../../../../reakit-test-utils/src/verify";
```
Comment (diegohaz) [inline packages/reakit/src/Box/__tests__/Box-test.tsx]: It should be something like this:
```suggestion
import { render, verify } from "reakit-test-utils";
```

File: packages/reakit/src/Box/__tests__/Box-test.tsx [hunk 3]
```diff
@@ -10,3 +11,10 @@ test("render", () => {
     </div>
   `);
 });
+
+test("has no a11y violations", async () => {
+  const { container } = render(<Box>box</Box>);
+  const results = await verify(container.innerHTML);
+
+  expect(results).toHaveNoViolations();
```
Comment (diegohaz) [inline packages/reakit/src/Box/__tests__/Box-test.tsx]: Nice! 👍 

I think we can do that for the example tests in follow-up PRs. Those have more real use cases.

File: packages/reakit-test-utils/src/verify.ts (hunk 1 above)
Comment (0ubbe) [reply packages/reakit-test-utils/src/verify.ts]: Cools, happy to change it to `axe` 👌🏻

**Comment by lluia** [comment]
## TODOs

- [ ] Bump `reakit-test-utils` with new `axe` utility → #654 
- [ ] Add **axe a11y checks** on existing unit tests on this PR

**Comment by lluia** [comment]
@diegohaz I was trying to add some **axe tests** but I encountered the following issues 🤔

- Some cleanup issues on `<Dialog />` tests ( _they pass fine withut the axe test_ ) 
Screenshot 2020-05-25 at 14 58 10

- Depends where I position the **axe test** one the component test file it'll corrupt other spapshot ( _probably related to some test cleanup problem_? )
Screenshot 2020-05-25 at 15 00 43

I don't have time today to dig deeper on this ⏰ ,  I'll have a look tomorrow. 
Posting in case you already know what might be the cause 👍🏻

**Comment by diegohaz** [comment]
Their docs say that the `cleanup` method should be called. I don't know.

Anyway, I think we should only check the examples tests.

**Comment by lluia** [comment]
Yup I tried with `cleanup()` as well and only in `Button/__examples__` but still seeing:
```
NotFoundError: The node to be removed is not a child of this node.
```
Tried to upgrade to `@testing-library/react@latest` with no better results though 😕 🤔

**Comment by lluia** [comment]
@diegohaz I was having a fresh look at this, and it seems it is related to `<Portal />`.

Passing 
```tsx
<Tooltip unstable_portal={false} />
``` 
on `<ButtonWithTooltip />` makes the `jest-axe` test pass 🤔 . I'll open an issue on `jest-axe` asking whether the library is compatible with `ReactDOM.createPortal` 👍🏻

**Comment by diegohaz** [comment]
@kennygrubb mentioned to use `baseElement` instead of `container` here: https://github.com/nickcolley/jest-axe/issues/105#issuecomment-649647708

And this actually makes total sense. `container` is only the `div` that RTL creates inside body. The portal element is actually added outside of it.

**Comment by lluia** [comment]
@diegohaz @kennygrubb coolios 💯  let me update this

**Comment by lluia** [comment]
Apologies @diegohaz 🙏🏽 , I finally got myself together to update this 👨🏽‍💻 , could you have a look? 

As suggested this was the ✨ formula that worked:
```js
await axe(baseElement)
``` 
I added the axe checks only on the examples tests as [per your suggestion](https://github.com/reakit/reakit/pull/652#discussion_r428318913).

I'm also happy to re-pick my work on the [`__examples__` story](https://github.com/reakit/reakit/issues/626) 👍🏽

**Comment by diegohaz** [comment]
> Apologies @diegohaz 🙏🏽 , I finally got myself together to update this 👨🏽‍💻

@lluia No worries! [Don't ever feel bad for not contributing to open source](https://github.com/reakit/reakit/blob/master/CONTRIBUTING.md#our-expectations-on-you-as-a-contributor). ❤️

**Pull request merged**